scylla/src/transport/timestamp_generator.rs

```diff
@@ -0,0 +1,100 @@
+use std::{
+    sync::atomic::AtomicI64,
+    time::{SystemTime, UNIX_EPOCH},
+};
+
+use async_trait::async_trait;
+use futures::lock::Mutex;
+use std::sync::atomic::Ordering;
+use tokio::time::{Duration, Instant};
+use tracing::warn;
+
+/// Trait used to represent a timestamp generator
+#[async_trait]
+pub trait TimestampGenerator: Send + Sync {
+    // This generates a new timestamp
+    async fn next_timestamp(&self) -> i64;
+}
+
+/// Basic monotonic timestamp generator
+pub struct MonotonicTimestampGenerator {
+    last: AtomicI64,
+    last_warning: Mutex<Instant>,
+    warning_threshold_us: i64,
+    warning_interval_ms: i64,
+}
```

Comment:
Currently you use raw integers for those durations, and you treat "0" as a special value.
The less error prone and more Rusty way is to use `std::time::Duration` to store thresholds / intervals, and use `Option<std::time::Duration>` instead of using a special value.